### ISSUE TITLE
zcash_client_backend: Make it possible to skip trial-decryption with internal keys.

### ISSUE THREAD

**Issue by nuttycom** [zcash/librustzcash#956]
Since internal keys should only ever receive funds that are being sent by the wallet, it should be possible to skip trial-decryption using the internal IVK for any block except those where a spend (either of a shielded note or a transparent UTXO) is detected.

**Comment by str4d** [comment]
This doesn't work reliably now that we are doing non-linear scanning; we don't detect spends for reverse scanning until after we've done trial decryption of the transactions they were spent in. We should still consider doing it where we do detect a spend during forward linear scanning.

**Comment by nuttycom** [comment]
I guess that we could start by just applying this optimization by checking whether the batch we're working on starts at the block above the fully-scanned height; it would give us a 2x improvement for our current approach to scanning big `Historic` ranges.

**Comment by daira** [comment]
> This doesn't work reliably now that we are doing non-linear scanning; we don't detect spends for reverse scanning until after we've done trial decryption of the transactions they were spent in.

Yes; the problem is that you might not know all of your notes at the point when you check a given transaction for whether it spends one of them. But you can put all of the nullifiers you see in a Bloom filter, and then whenever you find a note you will know whether it is possibly spent by a txn you previously saw, or definitely not spent by any of those txns. In the former case, you can go to secondary storage and do an expensive exact lookup of the nullifier. Writing the lookup table to secondary storage, e.g. as a [Log-Structured Merge tree](https://en.wikipedia.org/wiki/Log-structured_merge-tree), is mostly buffered sequential access so not too bad.
